src/librustc_typeck/outlives/explicit.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2013 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![allow(unused)]
+#[allow(dead_code)]
+
+use rustc::ty::{self, CratePredicatesMap, TyCtxt};
+use rustc::ty::maps::Providers;
+use std::rc::Rc;
+use util::nodemap::FxHashMap;
+use hir::map as hir_map;
+use rustc::hir;
+use rustc::hir::itemlikevisit::ItemLikeVisitor;
+use rustc::hir::def_id::{CrateNum, DefId, LOCAL_CRATE, LocalDefId};
+
+pub fn explicit_map<'tcx>(tcx: TyCtxt<'tcx, 'tcx, 'tcx>, crate_num: CrateNum)
+        -> CratePredicatesMap<'tcx> {
+
+    assert_eq!(crate_num, LOCAL_CRATE);
+    let empty_predicate = Rc::new(Vec::new());
+    let mut predicates: FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>> = FxHashMap();
+
+    {
+        let mut visitor = ExplicitVisitor {
+            tcx: tcx,
+            explicit_predicates: &mut predicates,
+            crate_num: crate_num,
+        };
+
+        //iterate over all the crates
+        tcx.hir.krate().visit_all_item_likes(&mut visitor);
+    }
+
+    ty::CratePredicatesMap {
+        predicates,
+        empty_predicate,
+    }
+}
+
+
+pub struct ExplicitVisitor<'tcx, 'a: 'tcx, 'p: 'a> {
+    tcx: TyCtxt<'tcx, 'tcx, 'tcx>,
+    explicit_predicates: &'a FxHashMap<DefId, Rc<Vec<ty::Predicate<'p>>>>,
+    crate_num: CrateNum,
+}
+
+impl<'tcx, 'a, 'p, 'v> ItemLikeVisitor<'v> for ExplicitVisitor<'tcx, 'a, 'p> {
```

Comment:
@nikomatsakis I am getting lifetime errors but not sure how to fix this.

```rust
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter `'tcx` due to conflicting requirements
  --> src/librustc_typeck/outlives/explicit.rs:59:49
   |
59 |         let local_explicit_predicate = self.tcx.explicit_predicates_of(def_id);
   |                                                 ^^^^^^^^^^^^^^^^^^^^^^
   |
note: first, the lifetime cannot outlive the lifetime 'tcx as defined on the impl at 54:1...
  --> src/librustc_typeck/outlives/explicit.rs:54:1
   |
54 | / impl<'tcx, 'a, 'p, 'v> ItemLikeVisitor<'v> for ExplicitVisitor<'tcx, 'a, 'p> {
55 | |     fn visit_item(&mut self, item: &hir::Item) {
56 | |
57 | |         //let def_id = LocalDefId::new(item.hir_id.owner).to_def_id();
...  |
77 | |     }
78 | | }
   | |_^
note: ...so that types are compatible (expected rustc::ty::TyCtxt<'_, '_, '_>, found rustc::ty::TyCtxt<'tcx, 'tcx, 'tcx>)
  --> src/librustc_typeck/outlives/explicit.rs:59:49
   |
59 |         let local_explicit_predicate = self.tcx.explicit_predicates_of(def_id);
   |                                                 ^^^^^^^^^^^^^^^^^^^^^^
note: but, the lifetime must be valid for the lifetime 'p as defined on the impl at 54:1...
  --> src/librustc_typeck/outlives/explicit.rs:54:1
   |
54 | / impl<'tcx, 'a, 'p, 'v> ItemLikeVisitor<'v> for ExplicitVisitor<'tcx, 'a, 'p> {
55 | |     fn visit_item(&mut self, item: &hir::Item) {
56 | |
57 | |         //let def_id = LocalDefId::new(item.hir_id.owner).to_def_id();
...  |
77 | |     }
78 | | }
   | |_^
note: ...so that expression is assignable (expected std::rc::Rc<std::vec::Vec<rustc::ty::Predicate<'p>>>, found std::rc::Rc<std::vec::Vec<rustc::ty::Predicate<'_>>>)
  --> src/librustc_typeck/outlives/explicit.rs:70:49
   |
70 |         self.explicit_predicates.insert(def_id, Rc::new(filtered_predicates));
   |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```